verification/tpm.go

```diff
@@ -0,0 +1,268 @@
+	var ec tpm2.EllipticCurve
+	var alg tpm2.Algorithm
+
+	profile, err := GetTransportProfile(d)
+	if err != nil {
+		t.Fatalf("Could not get profile: %v", err)
+	}
+	digestLen := profile.GetDigestSize()
+
+	if digestLen == len(SHA256Digest{0}) {
+		alg = tpm2.AlgSHA256
+		ec = tpm2.CurveNISTP256
+	} else if digestLen == len(SHA384Digest{0}) {
+		alg = tpm2.AlgSHA384
+		ec = tpm2.CurveNISTP384
+	}
+
+	//Create tpm auth session to get nonce and form label which is digest
+	tpm, err := tpm2.OpenTPM(*tpmPath)
+	if err != nil {
+		t.Fatalf("[FATAL]: Can't open TPM %s: %v", *tpmPath, err)
+	}
+
+	sessHandle, nonce, err := startTpmSession(t, tpm, alg)
+	if err != nil {
+		t.Fatalf("[FATAL]: Error in getting tpm nonce")
+	}
+
+	defer func() {
+		if err := tpm.Close(); err != nil {
+			t.Fatalf("[FATAL]: Can't close TPM %s: %v", *tpmPath, err)
+		}
+	}()
+
+	defer tpm2.FlushContext(tpm, sessHandle)
+
+	// Build SignHash request
+	expiry := int32(math.MinInt32)
+	digest := getDigest(nonce, expiry, digestLen)
+
+	seqLabel := make([]byte, digestLen)
+	for i, _ := range seqLabel {
+		seqLabel[i] = byte(i)
+	}
+
+	signParams := struct {
+		Flags      SignFlags
+		Label      []byte
+		ToBeSigned []byte
+	}{
+		Flags:      SignFlags(0),
+		Label:      seqLabel,
+		ToBeSigned: digest,
+	}
+
+	// Get signed hash from DPE
+	signResp, err := c.Sign(&ctx, signParams.Label, signParams.Flags, signParams.ToBeSigned)
+	if err != nil {
+		t.Fatalf("[FATAL]: Could not sign: %v", err)
+	}
+
+	// Get certificate chain of signing key
+	certifyKeyParams := struct {
+		Label []byte
+		Flags CertifyKeyFlags
+	}{
+		Label: seqLabel,
+		Flags: CertifyKeyFlags(0),
+	}
+
+	certifyKeyResp, err := c.CertifyKey(&(signResp.Handle), certifyKeyParams.Label, CertifyKeyX509, CertifyKeyFlags(0))
+	if err != nil {
+		t.Fatalf("[FATAL]: Could not CertifyKey: %v", err)
+	}
+
+	pubKey := extractPubKey(t, certifyKeyResp.Certificate)
+
+	// Get TPM handle loaded with public key
+	pkh := getPubKeyHandle(t, pubKey, tpm, alg, ec)
+
+	// Get encoded signature from TPM
+	r := new(big.Int).SetBytes(signResp.HmacOrSignatureR)
+	s := new(big.Int).SetBytes(signResp.SignatureS)
+
+	encodedSignature := getEncodedSignature(t, r, s, alg)
+
+	// Verify Policy with Signature
+	_, _, err = tpm2.PolicySigned(tpm, pkh, sessHandle, nonce, nil, nil, expiry, encodedSignature)
+	if err != nil {
+		t.Fatalf("[FATAL]: PolicySigned() failed: %v", err)
+	}
+
+	t.Log("[LOG]: PolicySigned() call success")
+}
+
+func getDigest(nonce []byte, expiry int32, digestLen int) []byte {
+
+	expBytes := make([]byte, 4)
+	binary.BigEndian.PutUint32(expBytes, uint32(expiry))
+
+	toDigest := append(nonce, expBytes...)
+
+	digest := make([]byte, digestLen)
+	if digestLen == len(SHA256Digest{0}) {
+		hash := sha256.Sum256(toDigest)
+		digest = hash[:]
+	} else if digestLen == len(SHA384Digest{0}) {
+		hash := sha512.Sum384(toDigest)
+		digest = hash[:]
+	}
+	return digest
+}
+
+func extractPubKey(t *testing.T, leafBytes []byte) any {
```

Comment:
I think ideally we would return an `ecdsa.PublicKey` here. If `pubKey.(type) != *ecdsa.PublicKey` we can fail.

Reply:
Validation is added and only ecdsa type pub key usd for further validation